components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,839 @@
+    // Last id of the pending task
+    last_unpersisted_id: u64,
+    // Last id of the known persisted task
+    last_persisted_id: u64,
+    // Used for getting the latest persisted id from sync worker
+    global_persisted_id: Arc<CachePadded<AtomicU64>>,
+    sync_sender: Sender<SyncMsg>,
+    raft_write_size_limit: usize,
+    metrics: StoreWriteMetrics,
+    message_metrics: RaftSendMessageMetrics,
+    perf_context: EK::PerfContext,
+}
+
+impl<EK, ER, T, N> Worker<EK, ER, T, N>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: Transport,
+    N: Notifier,
+{
+    fn new(
+        store_id: u64,
+        tag: String,
+        worker_id: usize,
+        engines: Engines<EK, ER>,
+        receiver: Receiver<WriteMsg<EK, ER>>,
+        notifier: N,
+        trans: T,
+        global_persisted_id: Arc<CachePadded<AtomicU64>>,
+        sync_sender: Sender<SyncMsg>,
+        config: &Config,
+    ) -> Self {
+        let batch = WriteTaskBatch::new(
+            engines.kv.write_batch_with_cap(KV_WB_DEFAULT_SIZE),
+            engines.raft.log_batch(RAFT_WB_DEFAULT_SIZE),
+        );
+        let perf_context = engines
+            .kv
+            .get_perf_context(config.perf_level, PerfContextKind::RaftstoreStore);
+        Self {
+            store_id,
+            tag,
+            worker_id,
+            engines,
+            receiver,
+            notifier,
+            trans,
+            batch,
+            pending_tasks: VecDeque::new(),
+            last_unpersisted_id: 0,
+            last_persisted_id: 0,
+            global_persisted_id,
+            sync_sender,
+            raft_write_size_limit: config.raft_write_size_limit.0 as usize,
+            metrics: StoreWriteMetrics::new(config.store_waterfall_metrics),
+            message_metrics: Default::default(),
+            perf_context,
+        }
+    }
+
+    fn run(&mut self) {
+        let mut notify_ready_buffer = HashMap::default();
+        loop {
+            let loop_begin = Instant::now();
+            let mut handle_begin = loop_begin;
+
+            let mut after_persist = Duration::from_micros(0);
+            let mut first_time = true;
+            while self.batch.get_raft_size() < self.raft_write_size_limit {
+                let msg = if first_time {
+                    first_time = false;
+                    match self.receiver.recv() {
+                        Ok(msg) => {
+                            self.metrics
+                                .task_gen
+                                .observe(duration_to_sec(loop_begin.saturating_elapsed()));
+                            handle_begin = Instant::now();
+                            msg
+                        }
+                        Err(_) => return,
+                    }
+                } else {
+                    match self.receiver.try_recv() {
+                        Ok(msg) => msg,
+                        Err(TryRecvError::Empty) => break,
```

Comment:
Should we need to block the worker if there is not any pending write.

Reply:
It will. Next time it will block on the `recv` function if no new msg.